### PR TITLE
Migrate to PureScript 0.14

### PR DESCRIPTION
**Pull request by fsoikin**
This PR makes necessary changes to build with the shiny new PureScript 0.14.0 compiler. There were just a few kinds of changes:

* `generics-rep` is not a thing anymore, it's been rolled into `prelude`.
* `Data.Generic.Show` and `Data.Generic.Eq` are now `Data.Show.Generic` and `Data.Eq.Generic` respectively.
* `Data.Either.fromRight` changed from `Either a b -> b` to `b -> Either a b -> b`. Replaced its use sites with `fromJust <<< hush`.

### CODE REVIEW

**Comment by thomashoneyman** [comment]
I was just taking a look to see if there was a 0.14 PR ready for this library. Glad to see it!

I suspect that this will be merged much faster if it only includes changes to the source code and updated major dependencies in the bower.json file (all that’s strictly necessary to update). The bower file is still needed to publish libraries to Pursuit at the moment — hopefully not for much longer.

**Comment by fsoikin** [comment]
Bot doesn't the bower file get generated by Spago now?

**Comment by thomashoneyman** [comment]
It can be, for libraries that are primarily spago-based. But for maintainers who don’t want to switch until the new registry is released, the bower file is still maintained using the bower cli (or by hand).

**Comment by fsoikin** [comment]
So like I said in https://github.com/cdepillabout/purescript-email-validate/pull/17, I tried with Bower, and it spits out a _ton_ of conflicts. My hunch is that this is because `ansi@5.0.0` (which is in the current `psc-0.14.0` package set) still [depends on `strings@4.0.0`](https://pursuit.purescript.org/packages/purescript-ansi/5.0.0), and that ripples out to everything else.

I'm trying to figure this out, but I don't really understand the mysterious ways of Bower...

**Comment by fsoikin** [comment]
Ok, @thomashoneyman, so I've made it work via Bower, but it's resolutions galore! 

* I upgraded all packages that were under "dependencies" to whatever version they have in the current `psc-0.14.0` package set.
* But one of them - `purescript-ansi` - still has old version `5.0.0`, which has three other old dependencies - `strings@4.0.0`, `foldable-traversable@4.0.0`, and `lists@5.0.0`.
* Of those three, only `lists` wasn't in the Bower file before, so I added it as `lists@6.0.0`.
* Still, somehow Bower managed to find a whole lot of outdated packages. Perhaps a problem similar to `ansi` exists somewhere in second- or third-tier dependencies?
* So to make it work I added all those resolutions.
* Still, even after that, the `free` package managed to end up outdated. No idea how. The way I noticed it was that it failed to compile.
* So I fixed that by pinning `free`'s version under "dependencies" too.

So now it compiles, but it feels supremely flimsy to me. I mean, if `free` managed to slip through, other packages could too, right? And I only noticed `free` because it failed to compile, so there could be silent failures.

**Comment by thomashoneyman** [comment]
Oof — you’ll need to update ansi, too, as you’re right its bower file is out of date and it will need a new release. It works in the package sets but unfortunately we do still have to support the bower workflows. I’m with you — bower is awful for this kind of thing — but it’s what we have to do for now. It often comes down to finding the transitive dependency that wasn’t updated.

**Comment by fsoikin** [comment]
Opened https://github.com/hdgarrood/purescript-ansi/pull/12

**Comment by fsoikin** [comment]
But it's not enough. Even pointing `purescript-ansi` to my branch in `bower.json` doesn't get rid of dependencies. 
Investigating further...

You wouldn't, by any chance, know a way to ask Bower which path of transitive dependencies leads to a particular one, a la `npm why`? A brief search didn't show any results :-/

**Comment by thomashoneyman** [comment]
All the libraries in the bower file are core or contrib libraries except for pipes and ansi. Core, contrib, web, and node libraries are all guaranteed updated for 0.14, so I suspect it’s the pipes library.

**Comment by thomashoneyman** [comment]
I don’t know of any command like that, unfortunately, though I’m no bower expert. It would help!

**Comment by JordanMartinez** [comment]
It's the `pipes` library. This is where I got to before I stopped with my fork when I was working on my learning repo.

**Comment by fsoikin** [comment]
@JordanMartinez I just found `pipes` too and working on it, but it goes deeper still: `pipes` depends on `mmorph`, which is itself outdated. But that seems to be the last one.

File: bower.json [hunk 1]
```diff
@@ -15,18 +15,17 @@
...
+    "purescript-avar": "^4.0.0",
+    "purescript-console": "^5.0.0",
+    "purescript-aff": "^6.0.0",
+    "purescript-exceptions": "^5.0.0",
+    "purescript-strings": "^5.0.0",
+    "purescript-prelude": "^5.0.0",
+    "purescript-transformers": "^5.0.0",
+    "purescript-foldable-traversable": "^5.0.0",
+    "purescript-pipes": "git://github.com/fsoikin/purescript-pipes#purescript-0.14",
+    "purescript-ansi": "git://github.com/fsoikin/purescript-ansi#purescript-0.14",
```
Comment (fsoikin) [inline bower.json:27]: These temporarily point to https://github.com/hdgarrood/purescript-ansi/pull/12 and https://github.com/felixSchl/purescript-pipes/pull/18

**Comment by fsoikin** [comment]
Success! Updating `ansi`, `pipes`, and `mmorph` did the trick!

* https://github.com/hdgarrood/purescript-ansi/pull/12
* https://github.com/felixSchl/purescript-pipes/pull/18
* https://github.com/hdgarrood/purescript-ansi/pull/12

File: bower.json [hunk 2]
```diff
@@ -15,18 +15,17 @@
...
-    "purescript-now": "^4.0.0"
+    "purescript-avar": "^4.0.0",
+    "purescript-console": "^5.0.0",
+    "purescript-aff": "^6.0.0",
+    "purescript-exceptions": "^5.0.0",
+    "purescript-strings": "^5.0.0",
+    "purescript-prelude": "^5.0.0",
+    "purescript-transformers": "^5.0.0",
+    "purescript-foldable-traversable": "^5.0.0",
+    "purescript-pipes": "git://github.com/fsoikin/purescript-pipes#purescript-0.14",
```
Comment (fsoikin) [inline bower.json:26]:
```suggestion
    "purescript-pipes": "v7.0.0",
```

**Comment by fsoikin** [comment]
Alright @thomashoneyman everything seems to be resolved: `ansi@6.0.0` and `pipes@7.0.0`

**Comment by thomashoneyman** [comment]
All looks good to me 👍 I’m not a maintainer, though, so I’m not able to merge or make a release.

**Comment by owickstrom** [comment]
> All looks good to me +1 I’m not a maintainer, though, so I’m not able to merge or make a release.

You are now! :+1: Feel free to make a new release if that works.

**Comment by fsoikin** [comment]
Oh good heavens! I'm a maintainer too!

image

**Comment by fsoikin** [comment]
https://pursuit.purescript.org/packages/purescript-spec/5.0.0

**Comment by owickstrom** [comment]
@fsoikin enjoy :smile:

**Pull request merged**